File: truss-compiler/src/main/java/truss/compiler/support/Arguments.java
package truss.compiler.support;

import truss.compiler.ArgumentException;
import org.apache.commons.lang.Validate;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Arguments {
    private final List<String> remaining = new ArrayList<>();
    private final Map<String, Argument> arguments = new HashMap<>();

    public void addArgument(Argument argument) {
        Validate.notNull(argument, "argument");

        Validate.isTrue(!arguments.containsKey(argument.getOption()));

        arguments.put(argument.getOption(), argument);
    }

    public void parse(String[] args) throws ArgumentException {
        Argument parsing = null;

        for (String arg : args) {
            if (parsing != null) {
                parsing.setProvided(true);
                parsing.setValue(arg);
                parsing = null;
            } else {
                parsing = arguments.get(arg);

                if (parsing == null) {
                    remaining.add(arg);
                } else {
                    if (!parsing.hasParameter()) {
                        parsing.setProvided(true);
                        parsing = null;
                    } else if (parsing.isProvided() && !parsing.allowMultiple()) {
                        throw new ArgumentException(String.format("Option '%s' can only appear once", arg));
                    }
                }
            }
        }

        for (Argument argument : arguments.values()) {
            if (argument.isMandatory() && !argument.isProvided()) {
                throw new ArgumentException(String.format("Option '%s' is mandatory", argument.getOption()));
            }
        }
    }

    public List<String> getRemaining() {
        return remaining;
    }
}

File: Truss.Compiler/Support/Arguments.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Truss.Compiler.Support {
    public class Arguments {
        private readonly Dictionary<string, Argument> _arguments = new Dictionary<string, Argument>();

        public Arguments() {
            Remaining = new List<string>();
        }

        public void AddArgument(Argument argument) {
            if (argument == null) {
                throw new ArgumentNullException("argument");
            }
            if (_arguments.ContainsKey(argument.Option)) {
                throw new ArgumentException("Duplicate argument");
            }

            _arguments.Add(argument.Option, argument);
        }

        public void Parse(string[] args) {
            Argument parsing = null;

            foreach (string arg in args) {
                if (parsing != null) {
                    parsing.IsProvided = true;
                    parsing.SetValue(arg);
                    parsing = null;
                } else {
                    parsing = _arguments[arg];

                    if (parsing == null) {
                        Remaining.Add(arg);
                    } else {
                        if (!parsing.HasParameter) {
                            parsing.IsProvided = true;
                            parsing = null;
                        } else if (parsing.IsProvided && !parsing.AllowMultiple) {
                            throw new ArgumentException(string.Format("Option '{0}' can only appear once", arg));
                        }
                    }
                }
            }

            foreach (Argument argument in _arguments.Values) {
                if (argument.IsMandatory && !argument.IsProvided) {
                    throw new ArgumentException(string.Format("Option '{0}' is mandatory", argument.Option));
                }
            }
        }

        public List<string> Remaining { get; private set; }
    }
}
